plutus-core/plutus-ir/src/PlutusIR/Transform/NewLetFloat.hs

```diff
@@ -0,0 +1,493 @@
+           => Marks
+           -> term
+           -> (FloatTable tyname name uni fun a, -- the floatable lets
+              term) -- the cleaned-up, reducted term
+removeLets marks = go
+  where
+    go :: term -> (FloatTable tyname name uni fun a, term)
+    go = breakNonRec >>> \case
+        -- main operation: for letrec or single letnonrec
+        Let a r bs@(representativeBindingUnique -> letU) tIn ->
+            let
+                -- go to rhs'es and collect their floattable + cleanedterm
+                (fRhs, bs') = M.unionsWith (<>) `first`
+                                 NE.unzip (goBinding <$> bs)
+                -- go to inTerm and collect its floattable + cleanedterm
+                (fIn, tIn') = go tIn
+                fBoth = M.unionWith (<>) fRhs fIn
+            in case marks M.!? letU  of
+                -- this is not a floatable let
+                -- propagate the floattable and KEEP this let in the cleaned term
+                Nothing  -> (fBoth, Let a r bs' tIn')
+                -- floatable let found.
+                -- move this let to the floattable, and just return the traversed interm
+                Just pos -> (M.insertWith (<>) pos (pure $ LetNaked a r bs') fBoth, tIn')
+
+        -- descend and collect
+        Apply a t1 t2 ->
+            let (f1, t1') = go t1
+                (f2, t2') = go t2
+            in (M.unionWith (<>) f1 f2, Apply a t1' t2')
+        TyInst a t ty -> flip (TyInst a) ty `second` go t
+        TyAbs a tyname k t -> TyAbs a tyname k `second` go t
+        LamAbs a name ty t -> LamAbs a name ty `second` go t
+        IWrap a ty1 ty2 t -> IWrap a ty1 ty2 `second` go t
+        Unwrap a t -> Unwrap a `second` go t
+
+        -- no term inside here, nothing to do
+        t@Var{} -> (mempty, t)
+        t@Constant{} -> (mempty, t)
+        t@Builtin{} -> (mempty, t)
+        t@Error{} -> (mempty, t)
+
+    goBinding :: Binding tyname name uni fun a
+              -> (FloatTable tyname name uni fun a,
+                 Binding tyname name uni fun a)
+    goBinding = \case
+        TermBind x s d t -> TermBind x s d `second` go t
+        -- no term inside here, nothing to do
+        b@TypeBind{}     -> (mempty, b)
+        b@DatatypeBind{} -> (mempty, b)
+
+-- | The 3rd and last pass that, given the result of 'removeLets', places the lets back (floats) at the right marked positions.
+floatBackLets :: forall tyname name uni fun a term re.
+                ( term~Term tyname name uni fun a
+                , re~Reader Depth
+                , PLC.HasUnique tyname PLC.TypeUnique, PLC.HasUnique name PLC.TermUnique)
+              => FloatTable tyname name uni fun a -- the lets to be floated
+              -> term -- the cleanedup, reducted term
+              -> term -- the final, floated, and correctly-scoped term
+floatBackLets fTable =
+    -- our reader context is only the depth this time.
+    flip runReader topDepth . goTop
+  where
+
+    goTop, go :: term -> re term
```

Comment:
This `re` thing is weird. If you're going to have a monad type variable, call it `m`!